### PR TITLE
[JavaScript] Add array-transformations Concept Exercise

### LINKED ISSUE #1896: [JavaScript] Implement new Concept Exercise: array-transformations
This issue describes how to implement the `array-transformations` concept exercise for the `javascript` track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

See the documentation _above_ (general documentation), as well as [How to implement a Concept Exercise in JavaScript](https://github.com/exercism/v3/blob/master/languages/javascript/reference/implementing-a-concept-exercise.md).

## Goal

The goal of this exercise is to teach the student how they can transform arrays.

## Learning objectives

- `.map`
- `.reduce`
- `.slice`
- `.splice`
- `.filter`
- `.sort`
- `.flatMap` (maybe)

## Out of scope

- anything that explicitly loops an array (array-loops)
- anything that returns a boolean (array-analysis)
- anything that implicitly loops (arrays)
  - `indexOf`
  - `find`
  - `findIndex`
  - `includes`
  - etc.

## Concepts

- `array-transformations`

## Prerequisites

- `arrays`

## Implementing

Use the story from `arrays` / `array-analysis`. 

## Help

You can choose to do this solo-style, or collaborate with multiple people on this. My suggestion would be is to 

1. **first accept this issue by saying "I'd like to work on this"** (you will not get a response from me. Just go with it) and optionally request that someone works with you (and wait for a second person to accept your request).
2. Use this issue to write down what you need (collect reference material) and discuss, if necessary, what to write and what not to write.
3. Draft a PR and request the feedback from at least one other JavaScript member (I recommend just pinging the channel on slack).

### PR DESCRIPTION
**Pull request by yyyc514**
Closes #1896

### CODE REVIEW

**Comment by yyyc514** [comment]
Not quite sure how nice samples should be:

```
export function twoIsSpecial(deck) {
   return deck.filter((card) => card === 2)
 }
```

**Comment by yyyc514** [comment]
Oh also I'm not sure if they introduction should talk about the different array methods more or if an example with just one of them is sufficient.

**Comment by yyyc514** [comment]
Any reason not to add `fill`? Be pretty easy to teach.

**Comment by SleeplessByte** [comment]
You can add fill!

**Comment by yyyc514** [comment]
Should we perhaps also depend on "callbacks" concept since so many of the array transforms require a callback function?

File: .editorconfig [hunk 1]
```diff
@@ -1,7 +1,10 @@
 root = true
 
 [*]
+end_of_line = lf
```
Comment (ErikSchierboom) [inline .editorconfig:4]: Was this meant to be in this PR? I saw the discussion in the #v3-markathon channel, and I'm not sure there was a decision made on this. I'd prefer this to be in a separate PR (if at all :)).

File: languages/javascript/exercises/concept/array-transformations/.docs/hints.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+## 1. Double every single card
+
+- There is a [built-in][map_method_docs] method great for simple transforms like this.
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/hints.md:3]: Would it perhaps make sense to also have the "method" text be linked to?

```suggestion
- There is a [built-in method][map_method_docs] great for simple transforms like this.
```

If you feel this is useful, this comment also applies to the other hints.

File: languages/javascript/exercises/concept/array-transformations/.docs/hints.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
+## 1. Double every single card
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/hints.md:1]: The hints are very nice! 👍

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
+Elyse, magician-to-be, continues her training. She will be given a deck of cards and attempt to transform that deck. She will make certain cards appear, disappear, change their values, or completely rearrange themselves.
+To make things easier, she usually only starts with cards numbered 1 to 10, although some of the tricks may result in larger cards being created.
+
+## 1. - Double every single card
+
+Elyse wants to double the number of every card in the deck. This may result in higher cards than 1-10.
+
+```javascript
+const deck = [1, 2, 3, 4, 10]
+seeingDouble(deck)
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:10]: I like the naming! :D

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
+Elyse, magician-to-be, continues her training. She will be given a deck of cards and attempt to transform that deck. She will make certain cards appear, disappear, change their values, or completely rearrange themselves.
+To make things easier, she usually only starts with cards numbered 1 to 10, although some of the tricks may result in larger cards being created.
+
+## 1. - Double every single card
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:4]:
```suggestion
## 1. Double every single card
```

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+Elyse wants to double the number of every card in the deck. This may result in higher cards than 1-10.
+
+```javascript
+const deck = [1, 2, 3, 4, 10]
+seeingDouble(deck)
+// => [2, 4, 6, 8, 20]
+```
+
+## 2. - Create multiple copies of every 3 found in the deck
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:14]:
```suggestion
## 2. Create multiple copies of every 3 found in the deck
```

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+Elyse wants to triplicate ever 3 found in the deck. If a deck started with a single 3, after the trick the deck would have three 3's in place of original.
+
+```javascript
+const deck = [1, 3, 9, 3, 7]
+threeOfEachThree(deck)
+// => [1, 3, 3, 3, 9, 3, 3, 3, 7]
+```
+
+## 3. - Find two cards from the exact middle of the deck
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:24]:
```suggestion
## 3. Find two cards from the exact middle of the deck
```

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+Elyse will take a deck and make every card disappear except the middle two cards. This trick is always performed with exactly 10 cards.
+
+```javascript
+const deck = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+middleTwo(deck)
+// => [5, 6]
+```
+
+## 4. - The outside two cards will reappear in the middle of the deck
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:34]:
```suggestion
## 4. The outside two cards will reappear in the middle of the deck
```

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+Elyse wants to move the top and bottom cards of the deck into the middle, in reverse order.
+
+```javascript
+const deck = [1, 2, 3, 5, 6, 10]
+sandwichTrick(deck)
+// => [2, 3, 10, 1, 5, 6]
+```
+
+## 5. - Every card that isn't 2 disappears
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:44]:
```suggestion
## 5. Every card that isn't 2 disappears
```

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+Elyse's favorite number today is 2. In this trick every card that isn't a 2 will disappear from the eck.
+
+```javascript
+const deck = [1, 2, 3, 4, 10, 2]
+twoIsSpecial(deck)
+// => [2, 2]
+```
+
+## 6. - Convert a shuffled deck into a perfectly ordered deck
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:54]:
```suggestion
## 6. Convert a shuffled deck into a perfectly ordered deck
```

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
+Elyse, magician-to-be, continues her training. She will be given a deck of cards and attempt to transform that deck. She will make certain cards appear, disappear, change their values, or completely rearrange themselves.
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:1]: I like how the existing theme is being re-used!

File: languages/javascript/exercises/concept/array-transformations/.docs/introduction.md [hunk 12]
```diff
@@ -0,0 +1,21 @@
+In Javascript, the Array class has many powerful built-in functions for transforming arrays. These functions make it much easier to do things than it otherwise would be using a simple for loop or more direct manipulation.
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/introduction.md:1]:
```suggestion
In Javascript, the `Array` class has many powerful built-in functions for transforming arrays. These functions make it much easier to do things than it otherwise would be using a simple for loop or more direct manipulation.
```

File: languages/javascript/exercises/concept/array-transformations/example.js [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * Removes every card from the deck except 2s
+ *
+ * @param {number[]} deck
+ *
+ * @returns {number[]} deck with only 2s
+ */
+export function twoIsSpecial(deck) {
+  return deck.filter((card) => card === 2)
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/example.js:64]: @yyyc514 Regarding your question how well-designed the example code should be, the general rule is to have the example be the most idiomatic solution you can write based on what the student has learnt in the exercise's prerequisites. In this particular case, I think in this case I would just have the magic numbers, but you could argue it either way. @SleeplessByte any preference?

File: languages/javascript/exercises/concept/array-transformations/.docs/after.md [hunk 14]
```diff
@@ -0,0 +1 @@
+Remember, anytime you're about to write a for loop to transform an array there is a good chance Array has already has you covered with a built-in.
```
Comment (ErikSchierboom) [inline languages/javascript/exercises/concept/array-transformations/.docs/after.md:1]: The `after.md` document should contain all the information that is in the `introduction.md` file. Part of the reason for this is that we'll make it easy for students to revisit the `after.md` document later, so it will be like a reference document for the involved concepts. The nice thing about the `after.md` document is that you can use it to expand upon things. For a list of things that you could include in the `after.md` document that you wouldn't include in the `introduction.md`, see https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsaftermd.

File: .editorconfig (hunk 1 above)
Comment (yyyc514) [reply .editorconfig:4]: I can pull it out and open a new PR for discussion.  It seems clear to me that our line endings for `*.md` files at LEAST should be consistent across the whole repository.  But discussion for another place.

File: languages/javascript/exercises/concept/array-transformations/.docs/hints.md (hunk 2 above)
Comment (yyyc514) [reply languages/javascript/exercises/concept/array-transformations/.docs/hints.md:3]: I was merely following the pattern established by other exercises. @SleeplessByte Should we change this everywhere?  I do see some merit.

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md (hunk 4 above)
Comment (yyyc514) [reply languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:10]: Thanks, I tried to make it light and fun.

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md (hunk 11 above)
Comment (yyyc514) [reply languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:1]: Just following the instructions. :)

File: languages/javascript/exercises/concept/array-transformations/.docs/after.md (hunk 14 above)
Comment (yyyc514) [reply languages/javascript/exercises/concept/array-transformations/.docs/after.md:1]: Yeah, this was more of a 'placeholder'... I'll watch the video again and read your link and then circle back to this.

File: languages/javascript/exercises/concept/array-transformations/.docs/hints.md (hunk 2 above)
Comment (yyyc514) [reply languages/javascript/exercises/concept/array-transformations/.docs/hints.md:3]: I actually don't love how we obscure the names of the methods￼ here. I feel like in any other context the name of the method would be part of the link text. ￼


Is that done on purpose because of the intention that these are hints?

File: languages/javascript/exercises/concept/array-transformations/example.js (hunk 13 above)
Comment (yyyc514) [reply languages/javascript/exercises/concept/array-transformations/example.js:64]: This might not of been the best example.  What about line 39 where at first glance the 4 and 5 have absolutely zero meaning to the reader. ￼

File: languages/javascript/exercises/concept/array-transformations/.docs/hints.md (hunk 2 above)
Comment (SleeplessByte) [reply languages/javascript/exercises/concept/array-transformations/.docs/hints.md:3]: The fact that there is a built-in method is the first hint. The name of the method is the implicit second hint. Giving away the name of the method removes the hit of endorphin someone would get if they choose _not_ to follow the link.

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+```javascript
+const deck = [1, 2, 3, 4, 10]
+seeingDouble(deck)
+// => [2, 4, 6, 8, 20]
+```
+
+## 2. - Create multiple copies of every 3 found in the deck
+
+Elyse wants to triplicate ever 3 found in the deck. If a deck started with a single 3, after the trick the deck would have three 3's in place of original.
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:16]:
```suggestion
Elyse wants to triplicate every 3 found in the deck. If a deck started with a single 3, after the trick the deck would have three 3's in place of original.
```

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+```javascript
+const deck = [1, 3, 9, 3, 7]
+threeOfEachThree(deck)
+// => [1, 3, 3, 3, 9, 3, 3, 3, 7]
+```
+
+## 3. - Find two cards from the exact middle of the deck
+
+Elyse will take a deck and make every card disappear except the middle two cards. This trick is always performed with exactly 10 cards.
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:26]:
```suggestion
Elyse will take a deck of ten cards, and make every card disappear except the middle two cards. 
```

File: languages/javascript/exercises/concept/array-transformations/.docs/instructions.md [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+
+```javascript
+const deck = [1, 2, 3, 5, 6, 10]
+sandwichTrick(deck)
+// => [2, 3, 10, 1, 5, 6]
+```
+
+## 5. - Every card that isn't 2 disappears
+
+Elyse's favorite number today is 2. In this trick every card that isn't a 2 will disappear from the eck.
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-transformations/.docs/instructions.md:46]:
```suggestion
Elyse's favorite number today is 2. In this trick every card that isn't a 2 will disappear from the deck.
```

File: languages/javascript/exercises/concept/array-transformations/example.js (hunk 13 above)
Comment (SleeplessByte) [reply languages/javascript/exercises/concept/array-transformations/example.js:64]: In reality, that one can be made more general purpose by removing the magic numbers.

```javascript
const middle = Math.floor(deck.length / 2)
return deck.slice(middle,  middle + 1)
```

File: .editorconfig (hunk 1 above)
Comment (ErikSchierboom) [reply .editorconfig]: > I can pull it out and open a new PR for discussion. It seems clear to me that our line endings for *.md files at LEAST should be consistent across the whole repository. But discussion for another place.

Seems like a plan!

File: languages/javascript/exercises/concept/array-transformations/example.js (hunk 13 above)
Comment (ErikSchierboom) [reply languages/javascript/exercises/concept/array-transformations/example.js:64]: This remains something of a delicate problem, where there is no definitive right or wrong in most cases. These are the rules according to [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#example-implementation-file):

- The implementation must be idiomatic.
- The code should be as simple as possible.
- Only use language features introduced by the exercise or its prerequisites (and their prerequisites, and so on).
- The example file is not shown to the student when doing in-browser coding and is not downloaded to the student's file system when using the CLI.

So as long as the example fits those rules, it is valid.

File: languages/javascript/exercises/concept/array-transformations/.docs/after.md (hunk 14 above)
Comment (ErikSchierboom) [reply languages/javascript/exercises/concept/array-transformations/.docs/after.md:1]: Okay great! We've been toying around with the idea of putting the literal text `TODO` in documents that need updating later. Something to consider doing if you'd like reviewers to ignore files when reviewing.

File: languages/javascript/exercises/concept/array-transformations/example.js (hunk 13 above)
Comment (SleeplessByte) [reply languages/javascript/exercises/concept/array-transformations/example.js:64]: > The code should be as simple as possible.

Idiomatic can definitely contradict with this, as is the case with this specific question.

Comment (ErikSchierboom) [reply languages/javascript/exercises/concept/array-transformations/example.js:64]: > Idiomatic can definitely contradict with this, as is the case with this specific question.

Correct. We should probably make the instructions a bit more clear.

File: languages/javascript/maintainers.md [hunk 18]
```diff
@@ -19,3 +19,4 @@ These awesome people help maintain the JavaScript track.
 - Ovidiu Miu (Slack: `@Ovidiu Miu`, GitHub: [ovidiu141](https://github.com/ovidiu141))
 - Pedro Romano (Slack: `@paparomeo`, GitHub: [paparomeo](https://github.com/paparomeo))
 - Rishi Kothari (Slack: `@Rishi Kothari`, GitHub: [rishiosaur](https://github.com/rishiosaur))
+- Josh Goebel (Slack: `@Josh Goebel`, GitHub: [yyyc514](https://github.com/yyyc514/))
```
[the remark this answers is not in the document]
Comment (SleeplessByte) [reply languages/javascript/maintainers.md:22]: If you can fix this one and the linter @yyyc514 , I'll merge it, and we can make subsequent edits on future PRs. Great work!!!

File: languages/javascript/exercises/concept/array-transformations/example.js (hunk 13 above)
Comment (yyyc514) [reply languages/javascript/exercises/concept/array-transformations/example.js:64]: So do we want the simple version of the more flexible one? :)  Someone let me know.

**Comment by yyyc514** [comment]
@SleeplessByte Ping.

**Comment by SleeplessByte** [comment]
Thank you @yyyc514. This is a great exercise <3

File: languages/javascript/exercises/concept/array-transformations/.docs/after.md [hunk 19]
```diff
@@ -0,0 +1,3 @@
+TODO: Add more.
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/array-transformations/.docs/after.md:1]: Note to self: create improvent issue
